lib/server.js

```diff
@@ -0,0 +1,131 @@
+var Docker = require('dockerode');
+var debug = require('debug')('docker-exec-websocket-server:lib:server');
+var through = require('through');
+var fs = require('fs');
+var ws = require('ws');
+var slugid = require('slugid');
+var _ = require('lodash');
+var path = require('path');
+var assert = require('assert');
+
+var dockerSocket = '/var/run/docker.sock';
+var stats = fs.statSync(dockerSocket);
+if (!stats.isSocket()) {
+  throw new Error('Are you sure the docker is running?');
+}
+var docker = new Docker({socketPath: dockerSocket});
+
+var dbgstream;
+if (process.argv[2] === '--dump') {
+  dbgstream = fs.createWriteStream(path.join(__dirname, '/dump.log'));
+}
+
+// Makes a new buffer with code number (ideally 0x00-0xFF)
+// appended to the start of the given buffer
+function muxBuffer (code, buffer) {
+  return Buffer.concat([new Buffer(String.fromCharCode(code)), buffer]);
+}
+
+/** Makes an instance of docker exec running command, then links the provided
+    websocket to the stream provided by docker exec
+
+    @param tty: boolean representing whether or not terminal emulation is active
+  */
+function execWithSocket (container, socket, command, tty) {
+  var options = {
+    AttachStdout: true,
+    AttachStderr: true,
+    AttachStdin: true,
+    Tty: tty,
+    Detach: false,
+    Cmd: command,
+  };
+  var attachoptions = {
+    stdin: true,
+    stdout: true,
+    stderr: true,
+    stream: true,
+  };
+  return container.exec(options, (err, exec) => {
+    if (err) { //err proof
+      throw err;
+    }
+    exec.start(attachoptions, (err, execStream) => {
+      if (err) { //err proof
+        throw err;
+      }
+      var strout = through((data) => {
+        if (dbgstream) {
+          dbgstream.write('OUT: ' + data + '\n-\n');
+        }
```

Comment:
What if we start... by writing, just before we reach `exec.start`:
```js
var dbgWrite = dbgstream ? (m, d) => dbgstream.write(m + ": " + d + "\n-\n") : () => {};
```
Then we don't need all these if statements... we just always output something to `dbgWrite`.
If `dbgstream` isn't set there is practically no overhead to calling an empty function with two strings that already exist...